File: engine_sim/carmode.py
"""Car mode as a reusable object, not a script.

``car.py`` is a terminal program: parse argv, loop, print a status line.  An
iPhone build cannot be that -- an iOS app is a view with a run loop, and it
needs to CREATE this thing, tick it, read it and stop it.  So the whole of car
mode lives here, and the CLI is a thin shell around it:

    car = CarMode(engine_key="rs3", telemetry=OBDTelemetry(...), rpm_map=...)
    car.start()
    ...  car.update(dt)  ...        # from a timer, a loop, whatever drives you
    car.stop()

Nothing in here knows how it is being driven, prints anything, or touches a
display.  That is the whole point: the same object serves the terminal, an iOS
app and the tests.
"""

from __future__ import annotations

import math

from . import presets
from .obd import RpmMap, ShiftDetector
from .simulator import Simulator
from .units import rpm_to_rads

# How hard the sim's crank is pulled toward the mapped rpm.  Fast enough to
# track a real engine, slow enough that one bad OBD sample cannot snap the
# pitch (the poll is 30-50 Hz and every sample is a fresh measurement).
_RPM_TRACK = 22.0
# With no data we sit at the engine's own idle rather than in silence, eased
# gently so a link dropout fades instead of stepping.
_IDLE_TRACK = 3.0


class CarMode:
    """Plays a chosen engine at a real car's rpm.  No UI, no I/O, no printing.

    Parameters
    ----------
    engine_key : preset key from ``presets.ALL``.
    telemetry  : anything with the Forza/OBD duck type -- ``rpm``, ``raw_rpm``,
                 ``throttle``, ``speed``, ``gear``, ``map_kpa``, ``baro_kpa``,
                 ``speed_valid`` and ``is_live()``.  Started and stopped by the
                 caller, not by us: on a phone the link outlives the screen.
    rpm_map    : an :class:`~engine_sim.obd.RpmMap`; defaults to a 1:1 map.
    synth_factory : ``f(sim) -> Synthesizer``, called on start and again on
                 every engine change.  It is a FACTORY rather than an instance
                 because a Synthesizer bakes its voicing tables from the engine
                 it was built with -- swapping cars means rebuilding it, the
                 same thing the GUI does.  ``None`` runs the data path with no
                 audio at all, which is how the link is diagnosed and how this
                 is tested.
    shift_pop  : cut the throttle when the gearbox shifts (the exhaust bang).
    """

    def __init__(self, engine_key="rs3", telemetry=None, rpm_map=None,
                 synth_factory=None, shift_pop=True):
        if engine_key not in presets.ALL:
            raise ValueError("unknown engine preset: %r" % (engine_key,))
        self.engine_key = engine_key
        self.sim = Simulator(presets.ALL[engine_key]())
        self.sim.ignition_on = True
        self.telemetry = telemetry
        self.rpm_map = rpm_map if rpm_map is not None else RpmMap("direct")
        self.synth_factory = synth_factory
        self.synth = None
        self.running = False
        self.shifter = ShiftDetector()
        self.shift_pop = bool(shift_pop)

        # live readouts, for whatever is displaying them (a status line, a
        # Toga label, a test)
        self.rpm_out = 0.0
        self.shifting = False
        self.label = next((l for k, l, _ in presets.PRESETS if k == engine_key),
                          engine_key)

    # ------------------------------------------------------------- lifecycle
    def start(self):
        if self.synth_factory is not None and self.synth is None:
            self.synth = self.synth_factory(self.sim)
            self.synth.start()
        self.running = True

    def stop(self):
        self.running = False
        if self.synth is not None:
            self.synth.stop()
            self.synth = None

    def set_engine(self, engine_key: str):
        """Swap the engine being played, keeping the link and the rpm map.

        The synthesizer is REBUILT, not rebound: it bakes its voicing tables
        (exhaust geometry, firing offsets, cavity reverbs) from the engine it
        was constructed with, so handing it a new simulator would keep playing
        the old car's timbre at the new car's rpm.  Crank speed carries across
        so the pitch does not jump on the way.
        """
        if engine_key not in presets.ALL:
            raise ValueError("unknown engine preset: %r" % (engine_key,))
        omega = self.sim.omega
        self.engine_key = engine_key
        self.sim = Simulator(presets.ALL[engine_key]())
        self.sim.ignition_on = True
        self.sim.omega = omega
        self.label = next((l for k, l, _ in presets.PRESETS if k == engine_key),
                          engine_key)
        if self.synth is not None:
            self.synth.stop()
            self.synth = None
        if self.running and self.synth_factory is not None:
            self.synth = self.synth_factory(self.sim)
            self.synth.start()

    # ------------------------------------------------------------------ tick
    def update(self, dt: float):
        """Advance one control step.  Call this as often as you like (100 Hz
        in the CLI); the synth integrates the crank at audio rate itself, so
        this only has to keep the targets fresh."""
        dt = min(max(dt, 0.0), 0.1)
        sim, eng = self.sim, self.sim.engine
        tm = self.telemetry

        if tm is not None and tm.is_live():
            self.rpm_map.observe(tm.raw_rpm, tm.throttle)
            self.rpm_out = self.rpm_map(tm.rpm, eng.idle_rpm, eng.redline_rpm)

            self.shifting = self.shift_pop and self.shifter.update(
                dt, tm.raw_rpm, tm.throttle, tm.speed)
            # an ignition cut IS a closed throttle as far as the exhaust is
            # concerned -- that is the bang
            sim.throttle = 0.0 if self.shifting else tm.throttle

            sim.omega += (rpm_to_rads(self.rpm_out) - sim.omega) * min(
                _RPM_TRACK * dt, 1.0)

            dtr = sim.drivetrain
            dtr.v = tm.speed if getattr(tm, "speed_valid", False) else 0.0
            dtr.gear = max(getattr(tm, "gear", 0), 0)
            if eng.induction != "na":
                map_kpa = getattr(tm, "map_kpa", 0.0)
                if map_kpa:      # your REAL boost drives the simulated compressor
                    sim.boost = max(0.0, (map_kpa - tm.baro_kpa) * 0.01)
                else:
                    sim._update_boost(dt)
        else:
            self.shifting = False
            sim.throttle = 0.0
            sim.omega += (rpm_to_rads(eng.idle_rpm) - sim.omega) * min(
                _IDLE_TRACK * dt, 1.0)
            self.rpm_out = sim.omega * 60.0 / (2.0 * math.pi)

        sim.crank_angle += sim.omega * dt
        return self.rpm_out

    # ---------------------------------------------------------------- status
    def status(self) -> dict:
        """Everything a display might want, already resolved."""
        tm = self.telemetry
        live = bool(tm is not None and tm.is_live())
        boost_bar = 0.0
        if tm is not None and getattr(tm, "map_kpa", 0.0):
            boost_bar = (tm.map_kpa - tm.baro_kpa) * 0.01
        return {
            "live": live,
            "link": ("LIVE" if live
                     else (getattr(tm, "status", "") or "waiting")),
            "car_rpm": getattr(tm, "raw_rpm", 0.0) if tm else 0.0,
            "sim_rpm": self.rpm_out,
            "pedal": getattr(tm, "throttle", 0.0) if tm else 0.0,
            "gear": getattr(tm, "gear", 0) if tm else 0,
            "speed_kmh": (getattr(tm, "speed", 0.0) * 3.6) if tm else 0.0,
            "boost_bar": boost_bar,
            "hz": getattr(tm, "hz", 0.0) if tm else 0.0,
            "rtt_ms": (getattr(tm, "rtt", 0.0) * 1000.0) if tm else 0.0,
            "shifting": self.shifting,
            "shifts": self.shifter.shifts,
            "engine": self.label,
        }

File: swift/EngineSimCore/Sources/EngineSimCore/CarMode.swift
//
//  CarMode.swift
//  Play a chosen engine at a real car's revs.
//
//  No UI, no I/O, no printing -- this is the control layer, and it is a
//  separate object from the app precisely so it can be driven by a test with
//  no dongle and no audio device.
//
//  The one design decision worth stating: swapping engines REBUILDS the
//  synthesizer rather than rebinding it.  A Synthesizer bakes its voicing
//  tables -- exhaust geometry, firing offsets, cavity reverbs -- from the
//  engine it was constructed with, so handing it a new preset would keep
//  playing the old car's timbre at the new car's revs.  Crank speed carries
//  across so the pitch does not jump on the way through.
//

import Foundation

/// Anything that can say what the car is doing.  A live ELM327 and a slider
/// both satisfy it, which is how the whole path is exercised on a desk.
public protocol TelemetrySource: AnyObject {
    var rpm: Double { get }          // projected forward; what you play
    var rawRPM: Double { get }       // un-projected; what shifts are found in
    var throttle: Double { get }
    var speed: Double { get }        // m/s
    var gear: Int { get }
    var mapKPa: Double { get }
    var baroKPa: Double { get }
    var speedValid: Bool { get }
    var hz: Double { get }
    var status: String { get }
    func isLive() -> Bool
}

extension OBDTelemetry: TelemetrySource {
    public func isLive() -> Bool { status == "live" && lastPacket > 0 }
}

/// A hand-driven source: the sliders, for setting the thing up without a car.
public final class ManualSource: TelemetrySource {
    public var rpm: Double = 800
    public var rawRPM: Double { rpm }
    public var throttle: Double = 0
    public var speed: Double = 0
    public var gear: Int = 0
    public var mapKPa: Double = 0
    public var baroKPa: Double = 101.3
    public var speedValid: Bool = true
    public var hz: Double = 0
    public var status: String = "manual"
    public init() {}
    public func isLive() -> Bool { true }
}

public struct CarStatus {
    public var live = false
    public var link = "waiting"
    public var carRPM = 0.0
    public var simRPM = 0.0
    public var pedal = 0.0
    public var gear = 0
    public var speedKmh = 0.0
    public var boostBar = 0.0
    public var hz = 0.0
    public var shifting = false
    public var engine = ""
}

public final class CarMode {
    /// How fast the simulated crank chases the mapped target.  Fast enough to
    /// feel connected, slow enough that a single bad OBD sample is a wobble
    /// rather than a click.
    static let rpmTrack = 14.0
    static let idleTrack = 3.0

    public private(set) var engineKey: String
    public private(set) var engine: EnginePreset
    public var telemetry: TelemetrySource?
    public var rpmMap: RpmMap
    public var shiftPop = true
    public private(set) var shifting = false
    public private(set) var rpmOut = 0.0
    public private(set) var running = false
    public private(set) var synth: Synthesizer?

    let shifter = ShiftDetector()
    let library: EngineLibrary
    /// Rebuilt on every engine change -- see the file comment.
    let makeSynth: (EnginePreset, EngineTables, VoicingSetup) -> Synthesizer

    var omega = 0.0       // rad/s
    var boost = 0.0

    public init(engineKey: String, library: EngineLibrary,
                telemetry: TelemetrySource? = nil,
                rpmMap: RpmMap = RpmMap(mode: .direct),
                makeSynth: @escaping (EnginePreset, EngineTables, VoicingSetup)
                    -> Synthesizer) throws {
        guard let e = library.engine(engineKey) else {
            throw NSError(domain: "CarMode", code: 1, userInfo: [
                NSLocalizedDescriptionKey: "unknown engine preset: \(engineKey)"])
        }
        self.engineKey = engineKey
        self.engine = e
        self.library = library
        self.telemetry = telemetry
        self.rpmMap = rpmMap
        self.makeSynth = makeSynth
        omega = e.idleRpm * 2.0 * Double.pi / 60.0
    }

    // ------------------------------------------------------------ lifecycle
    public func start() {
        if synth == nil, let t = library.tables(engineKey),
           let v = library.voicing(engineKey) {
            synth = makeSynth(engine, t, v)
        }
        running = true
    }

    public func stop() {
        running = false
        synth = nil
    }

    /// Swap the engine, keeping the link and the rpm map.
    public func setEngine(_ key: String) throws {
        guard let e = library.engine(key), let t = library.tables(key),
              let v = library.voicing(key) else {
            throw NSError(domain: "CarMode", code: 1, userInfo: [
                NSLocalizedDescriptionKey: "unknown engine preset: \(key)"])
        }
        engineKey = key
        engine = e
        synth = nil
        if running { synth = makeSynth(e, t, v) }   // rebuilt, never rebound
    }

    // ----------------------------------------------------------------- tick
    /// Advance one control step.  Call it as often as you like -- the synth
    /// integrates the crank at audio rate itself, so this only has to keep the
    /// targets fresh.
    @discardableResult
    public func update(dt rawDt: Double) -> Double {
        let dt = min(max(rawDt, 0.0), 0.1)
        let idleOmega = engine.idleRpm * 2.0 * Double.pi / 60.0

        if let tm = telemetry, tm.isLive() {
            rpmMap.observe(rpm: tm.rawRPM, pedal: tm.throttle)
            rpmOut = rpmMap(tm.rpm, engIdle: engine.idleRpm,
                            engRedline: engine.redlineRpm)

            shifting = shiftPop && shifter.update(dt: dt, rpm: tm.rawRPM,
                                                  pedal: tm.throttle,
                                                  speed: tm.speed)
            // an ignition cut IS a closed throttle as far as the exhaust is
            // concerned -- that is the bang
            let throttle = shifting ? 0.0 : tm.throttle

            let target = rpmOut * 2.0 * Double.pi / 60.0
            omega += (target - omega) * min(CarMode.rpmTrack * dt, 1.0)

            if engine.induction != "na" && tm.mapKPa > 0 {
                // YOUR real boost drives the simulated compressor
                boost = max(0.0, (tm.mapKPa - tm.baroKPa) * 0.01)
            }
            if let s = synth {
                s.set(rpm: rpmOut, throttle: throttle, boost: boost)
                s.speed = tm.speedValid ? tm.speed : 0.0
                s.drive.gear = max(tm.gear, 0)
                s.drive.speed = s.speed
            }
        } else {
            shifting = false
            omega += (idleOmega - omega) * min(CarMode.idleTrack * dt, 1.0)
            rpmOut = omega * 60.0 / (2.0 * Double.pi)
            synth?.set(rpm: rpmOut, throttle: 0.0, boost: 0.0)
        }
        return rpmOut
    }

    // --------------------------------------------------------------- status
    public func status() -> CarStatus {
        var s = CarStatus()
        let tm = telemetry
        s.live = tm?.isLive() ?? false
        s.link = s.live ? "LIVE" : (tm?.status.isEmpty == false
                                    ? tm!.status : "waiting")
        s.carRPM = tm?.rawRPM ?? 0
        s.simRPM = rpmOut
        s.pedal = tm?.throttle ?? 0
        s.gear = tm?.gear ?? 0
        s.speedKmh = (tm?.speed ?? 0) * 3.6
        if let t = tm, t.mapKPa > 0 { s.boostBar = (t.mapKPa - t.baroKPa) * 0.01 }
        s.hz = tm?.hz ?? 0
        s.shifting = shifting
        s.engine = engine.name
        return s
    }
}

/// The three fixture files, loaded once and looked up by preset key.
public final class EngineLibrary {
    let engines: [String: EnginePreset]
    let tablesByKey: [String: EngineTables]
    let voicings: [String: VoicingSetup]
    let torques: [String: TorqueTable]

    public init(presets: Data, tables: Data, voicing: Data,
                torque: Data? = nil) throws {
        engines = try PresetLibrary.load(jsonData: presets)
        tablesByKey = try EngineTables.load(jsonData: tables)
        voicings = try VoicingSetup.load(jsonData: voicing)
        torques = try torque.map { try TorqueTable.load(jsonData: $0) } ?? [:]
    }

    public func torque(_ k: String) -> TorqueTable? { torques[k] }

    public var keys: [String] { engines.keys.sorted() }
    public func engine(_ k: String) -> EnginePreset? { engines[k] }
    public func tables(_ k: String) -> EngineTables? { tablesByKey[k] }
    public func voicing(_ k: String) -> VoicingSetup? { voicings[k] }
    public func name(_ k: String) -> String { engines[k]?.name ?? k }
}
